File: prototypes/incremental/src/effects.lua
-- src/effects.lua
local effects = {}

local shakes    = {}
local particles = {}

local SHAKE_DURATION  = 0.25
local SHAKE_INTENSITY = 2.5

function effects.shake(building_id)
  shakes[building_id] = SHAKE_DURATION
end

function effects.get_shake_offset(building_id)
  if not shakes[building_id] then return 0, 0 end
  local t   = shakes[building_id] / SHAKE_DURATION
  local amp = SHAKE_INTENSITY * t
  return (math.random() * 2 - 1) * amp, (math.random() * 2 - 1) * amp
end

function effects.spawn_particles(x, y, count, r, g, b)
  for _ = 1, count do
    local angle = math.random() * math.pi * 2
    local speed = math.random(15, 45)
    local life  = 0.4 + math.random() * 0.3
    table.insert(particles, {
      x = x, y = y,
      vx = math.cos(angle) * speed,
      vy = math.sin(angle) * speed - 40,
      life = life, max_life = life,
      r = r, g = g, b = b,
    })
  end
end

function effects.update(dt)
  for id, t in pairs(shakes) do
    shakes[id] = t - dt
    if shakes[id] <= 0 then shakes[id] = nil end
  end
  for i = #particles, 1, -1 do
    local p = particles[i]
    p.life = p.life - dt
    if p.life <= 0 then
      table.remove(particles, i)
    else
      p.x  = p.x  + p.vx * dt
      p.y  = p.y  + p.vy * dt
      p.vy = p.vy + 120 * dt
    end
  end
end

function effects.draw_particles()
  for _, p in ipairs(particles) do
    local alpha = p.life / p.max_life
    love.graphics.setColor(p.r, p.g, p.b, alpha)
    love.graphics.rectangle("fill", p.x - 1, p.y - 1, 2, 2)
  end
end

return effects

File: prototypes/minesweeper/src/effects.js
export function createEffects() {
  let shakeIntensity = 0;
  const tileShakes = new Map(); // "tx,ty" -> { elapsed, duration }
  const particles  = [];

  function triggerScreenShake(intensity) {
    shakeIntensity = Math.max(shakeIntensity, intensity);
  }

  function triggerTileShake(tx, ty) {
    tileShakes.set(`${tx},${ty}`, { elapsed: 0, duration: 0.25 });
  }

  function emitParticles(worldX, worldY, color, count) {
    for (let i = 0; i < count; i++) {
      const angle = Math.random() * Math.PI * 2;
      const speed = 40 + Math.random() * 120;
      particles.push({
        x: worldX, y: worldY,
        vx: Math.cos(angle) * speed,
        vy: Math.sin(angle) * speed - 30,
        life: 1,
        decay: 2 + Math.random() * 2,
        size: 2 + Math.random() * 3,
        color,
      });
    }
  }

  function update(dt) {
    shakeIntensity *= Math.max(0, 1 - dt * 10);
    if (shakeIntensity < 0.2) shakeIntensity = 0;

    for (const [key, s] of tileShakes) {
      s.elapsed += dt;
      if (s.elapsed >= s.duration) tileShakes.delete(key);
    }

    for (let i = particles.length - 1; i >= 0; i--) {
      const p = particles[i];
      p.x  += p.vx * dt;
      p.y  += p.vy * dt;
      p.vy += 250 * dt; // gravity
      p.life -= p.decay * dt;
      if (p.life <= 0) particles.splice(i, 1);
    }
  }

  function getScreenShakeOffset() {
    if (shakeIntensity < 0.2) return { x: 0, y: 0 };
    const a = Math.random() * Math.PI * 2;
    return { x: Math.cos(a) * shakeIntensity, y: Math.sin(a) * shakeIntensity };
  }

  function getTileShakeOffset(tx, ty) {
    const s = tileShakes.get(`${tx},${ty}`);
    if (!s) return { x: 0, y: 0 };
    const amp = (1 - s.elapsed / s.duration) * 4;
    return {
      x: Math.sin(s.elapsed * 80) * amp,
      y: Math.cos(s.elapsed * 65) * amp,
    };
  }

  return { triggerScreenShake, triggerTileShake, emitParticles, update,
           getScreenShakeOffset, getTileShakeOffset, particles };
}
